codex/merkletree/merkletree.nim

```diff
@@ -0,0 +1,164 @@
+  getLowHigh(self.leavesCount, level)
+
+func getTotalSize(leavesCount: int): int =
+  let height = calcTreeHeight(leavesCount)
+  getLowHigh(leavesCount, height - 1)[1] + 1
+
+proc getWidth(self: MerkleTree, i: int): int =
+  let (low, high) = self.getLowHigh(i)
+  high - low + 1
+
+func getChildren(self: MerkleTree, i, j: int): (MerkleHash, MerkleHash) =
+  let (low, high) = self.getLowHigh(i - 1)
+  let leftIdx = low + 2 * j
+  let rightIdx = min(leftIdx + 1, high)
+
+  (self.nodes[leftIdx], self.nodes[rightIdx])
+
+func getSibling(self: MerkleTree, i, j: int): MerkleHash =
+  let (low, high) = self.getLowHigh(i)
+  if j mod 2 == 0:
+    self.nodes[min(low + j + 1, high)]
+  else:
+    self.nodes[low + j - 1]
+
+proc setNode(self: var MerkleTree, i, j: int, value: MerkleHash): void =
+  let (low, _) = self.getLowHigh(i)
+  self.nodes[low + j] = value
+
+proc root*(self: MerkleTree): MerkleHash =
+  self.nodes[^1]
+
+proc len*(self: MerkleTree): int =
+  self.nodes.len
+
+proc leaves*(self: MerkleTree): seq[MerkleHash] =
+  self.nodes[0..<self.leavesCount]
+
+proc nodes*(self: MerkleTree): seq[MerkleHash] =
+  self.nodes
+
+proc height*(self: MerkleTree): int =
+  calcTreeHeight(self.leavesCount)
+
+proc `$`*(self: MerkleTree): string =
+  result &= "leavesCount: " & $self.leavesCount
+  result &= "\nnodes: " & $self.nodes
+
+proc getProof*(self: MerkleTree, index: int): ?!MerkleProof =
+  if index > self.leaves.high or index < 0:
+    return failure("Index " & $index & " out of range [0.." & $self.leaves.high & "]" )
+
+  var path = newSeq[MerkleHash](self.height - 1)
+  for i in 0..<path.len:
+    let j = index div (1 shl i)
+    path[i] = self.getSibling(i, j)
+
+  success(MerkleProof(index: index, path: path))
+
+proc initTreeFromLeaves(leaves: seq[MerkleHash]): ?!MerkleTree =
+
+  without mcodec =? leaves.?[0].?mcodec and
+          digestSize =? leaves.?[0].?size:
+    return failure("At least one leaf is required")
+
+  if not leaves.allIt(it.mcodec == mcodec):
+    return failure("All leaves must use the same codec")
+
+  let totalSize = getTotalSize(leaves.len)
+  var tree = MerkleTree(leavesCount: leaves.len, nodes: newSeq[MerkleHash](totalSize))
+
+  var buf = newSeq[byte](digestSize * 2)
+  proc combine(l, r: MerkleHash): ?!MerkleHash =
+    copyMem(addr buf[0], unsafeAddr l.data.buffer[0], digestSize)
+    copyMem(addr buf[digestSize], unsafeAddr r.data.buffer[0], digestSize)
+
+    MultiHash.digest($mcodec, buf).mapErr(
+      c => newException(CatchableError, "Error calculating hash using codec " & $mcodec & ": " & $c)
+    )
+
+  # copy leaves
+  for j in 0..<leaves.len:
+    tree.setNode(0, j, leaves[j])
+
+  # calculate intermediate nodes
+  for i in 1..<tree.height:
+    for j in 0..<tree.getWidth(i):
+      let (left, right) = tree.getChildren(i, j)
+      
+      without mhash =? combine(left, right), error:
+        return failure(error)
+      tree.setNode(i, j, mhash)
```

Comment:
I like how there are many short functions. Very nice! But, there is quite a bit of `i` and `j` going on in these functions as well. Perhaps some of them can be renamed? index, level, or something else more descriptive?

Reply:
What do you think about this naming scheme: `level, i` instead of `i, j`? We would interpret that as an `i'th` node on a level `level`

Reply:
We had an offline conversation, and `level, i` seem to be good enough naming scheme.